src/nvim/eval.c

```diff
@@ -13131,16 +13155,13 @@ static void f_readfile(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     }
   }   /* while */
 
-  /*
-   * For a negative line count use only the lines at the end of the file,
-   * free the rest.
-   */
-  if (maxline < 0)
-    while (cnt > -maxline) {
-      tv_list_item_remove(rettv->vval.v_list,
-                          tv_list_first(rettv->vval.v_list));
-      cnt--;
-    }
+  // For a negative line count use only the lines at the end of the file,
+  // free the rest.
+  if (maxline < -tv_list_len(rettv->vval.v_list)) {
+    listitem_T *const first_li = tv_list_find(rettv->vval.v_list, maxline);
+    listitem_T *const last_li = tv_list_last(rettv->vval.v_list);
+    tv_list_remove_items(rettv->vval.v_list, first_li, last_li);
```

Comment:
Here the last items of the list are removed. In the old code the first items were removed.